docs/docs/tutorial/creating-a-source-plugin/part-3/index.mdx

```diff
@@ -0,0 +1,587 @@
+export { sourceNodes } from "./source-nodes"
+// highlight-next-line
+export { createSchemaCustomization } from "./create-schema-customization"
+```
+
+Also double check that you're running the `develop:deps` script in your terminal so that your changes to the plugin are compiled.
+
+### `createTypes`
+
+Before using [`createTypes`](/docs/reference/config-files/actions/#createTypes), here are some things to note for its usage:
+
+- Type definitions can be provided either in GraphQL's [schema definition language (SDL)](https://graphql.org/learn/schema/), Gatsby Type Builders, or a combination of both. You'll use the SDL syntax for the rest of this tutorial but you can also read an example for Gatsby's Type Builder syntax in this part.
+- By default, explicit type definitions you add with `createTypes` will be merged with inferred field types. So if you only want to define a subset of fields, the rest will be inferred as before. You can modify this behavior with the `@infer` and `@dontInfer` extensions. You'll learn more about extensions in [Add a foreign-key relationship](#add-a-foreign-key-relationship).
+- Type definitions targeting root node types, e.g. `MarkdownRemark` or others added in `sourceNodes`/`onCreateNode` like `Post` and `Author`, need to implement the `Node` interface. You can do this by adding `implements Node` to the SDL.
+
+Using an arbitrary example, let's imagine this is the shape of incoming data to `createNode` that gets inspected and inferred:
+
+```json
+{
+  "name": "Gatsby",
+  "address": {
+    "street": "Some street",
+    "city": "San Francisco",
+    "state": "California",
+  }
+}
+```
+
+In this example, let's pretend that you know `state` is not always provided — meaning that you want to explicitly define it to avoid the situation described above. Conversely, you also know that `street` will always be defined, and as such, you can mark it as [Non-Null](https://graphql.org/learn/schema/#lists-and-non-null) with a `!`. Lastly, `city` will be automatically inferred and merged with the custom types.
+
+```ts
+import type { GatsbyNode } from "gatsby"
+
+export const createSchemaCustomization: GatsbyNode[`createSchemaCustomization`] =
+  ({ actions }) => {
+    const { createTypes } = actions
+
+    createTypes(`
+      type User implements Node {
+        address: UserAddress
+      }
+
+      type UserAddress {
+        street: String!
+        state: String
+      }
+    `)
+  }
+```
+
+In this example the `User` type is a root node type, so you'll be able to query it with `allUser` and `user` in GraphQL.
+
+<Collapsible
+  summary={<em>Gatsby Type Builder example</em>}
+>
+
+In many cases, GraphQL SDL provides a succinct way to provide type definitions for your schema. If however you need more flexibility, `createTypes` also accepts type definitions provided with the help of Gatsby Type Builders, which are more flexible than SDL syntax. They are accessible on the `schema` argument passed to Node APIs.
```

Comment:
Not necesairly need to change anything, but they way I view SDL vs type builders is like this:

SDL is more terse, easier to read and to understand. Best choice for hardcoded schemas unless you need custom resolvers which force you to do something on top of SDL to provide those resolvers (either register custom directive or do both SDL and type builder / createResolvers). It becomes less easy to read if SDL is being generated through some kind of string concatenation/interpolation/manipulation.

Type Builders are better suited for cases when you can't "hardcode" schema and/or you need custom resolvers - if CMS has ability to create custom types with custom fields and you can fetch those definitions - it's easier to use typebuilders than generating SDL that would match it